src/slice-assign.h

```diff
@@ -0,0 +1,29 @@
+#ifndef VCTRS_SLICE_ASSIGN_H
+#define VCTRS_SLICE_ASSIGN_H
+
+struct vec_assign_opts {
+  bool assign_names;
+  struct vctrs_arg* x_arg;
+  struct vctrs_arg* value_arg;
+};
+
+static inline struct vec_assign_opts new_vec_assign_opts(bool assign_names,
+                                                         struct vctrs_arg* x_arg,
+                                                         struct vctrs_arg* value_arg) {
+  return (struct vec_assign_opts) {
+    .assign_names = assign_names,
+    .x_arg = x_arg,
+    .value_arg = value_arg
+  };
+}
+
+SEXP vec_assign_opts(SEXP x, SEXP index, SEXP value,
+                     const struct vec_assign_opts* opts);
+
+SEXP vec_proxy_assign_opts(SEXP proxy, SEXP index, SEXP value,
+                           const struct vec_assign_opts* opts);
+
+SEXP df_assign(SEXP x, SEXP index, SEXP value,
+               const struct vec_assign_opts* opts);
```

Comment:
Does it make sense to go ahead and move `chr_assign()`, `list_assign()` and `vec_assign()` here too?